### PR TITLE
Random cleanups

### PR DESCRIPTION
**Pull request by quad**
- Use iterators in `Newer::new`
- Drop unnecessary `.clone()`s in `find_newest`
- Use `Result<(), Error>` in `main`

Don't take any of this seriously.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -149,10 +119,10 @@ fn run_command(command: Vec<String>) -> Result<i32, Error> {
         .unwrap_or(-1))
 }
 
-fn find_newest(path: String) -> Result<SystemTime, Error> {
+fn find_newest(path: &str) -> Result<SystemTime, Error> {
```
Comment (quad) [inline src/main.rs:122]: `String` takes ownership; there doesn't seem to be a ready for that.

File: src/main.rs [hunk 2]
```diff
@@ -46,26 +46,14 @@ Use MK_LOG=trace to see debug output.
...
-            eprintln!("{}", e);
-            exit(1);
-        }
-    }
+    let newer = Newer::new(args)?;
+    let code = if newer.should_run_command() {
+        run_command(newer.command)?
+    } else {
+        debug!("Nothing to do.");
+        newer.newer.into()
```
Comment (quad) [inline src/main.rs:54]: I believe this does the same as `if newer.newer { 1 } else { 0 }`…

File: src/main.rs [hunk 3]
```diff
@@ -46,26 +46,14 @@ Use MK_LOG=trace to see debug output.
...
-            }
-        }
-        Err(e) => {
-            eprintln!("{}", e);
-            exit(1);
-        }
-    }
+    let newer = Newer::new(args)?;
+    let code = if newer.should_run_command() {
+        run_command(newer.command)?
```
Comment (quad) [inline src/main.rs:51]: **Behavior change**

In the original code, the exit code from `run_command` was swallowed and replaced by `0`.

File: src/main.rs [hunk 4]
```diff
@@ -75,53 +63,35 @@ struct Newer {
...
-        }
+        let mut args = args.iter();
+
+        let newest = args
+            .by_ref()
+            .take_while(|&a| a != ":")
+            .map(|arg| match find_newest(arg) {
+                Err(e) if e.kind() == ErrorKind::NotFound => Ok(SystemTime::UNIX_EPOCH),
+                n @ _ => n,
+            })
```
Comment (quad) [inline src/main.rs:74]: nit: a `.filter_map` is probably more appropriate here. Something like:

```rust
.filter_map(|arg| match find_newest(arg) {
    Err(e) if e.kind() == ErrorKind::NotFound => None,
    n @ _ => Some(n),
})
```

File: src/main.rs (hunk 2 above)
Comment (alecthomas) [reply src/main.rs:54]: Ah neat.

File: src/main.rs (hunk 3 above)
Comment (alecthomas) [reply src/main.rs:51]: Ah good catch, fixed that in main.

File: src/main.rs (hunk 4 above)
Comment (alecthomas) [reply src/main.rs:74]: Definitely not a fan of using iterators for this, though it's interesting.

File: src/main.rs (hunk 1 above)
Comment (alecthomas) [reply src/main.rs:122]: Good point, fixed in main!

File: src/main.rs [hunk 5]
```diff
@@ -75,53 +63,35 @@ struct Newer {
...
-            Command,
-        }
+        let mut args = args.iter();
+
+        let newest = args
+            .by_ref()
+            .take_while(|&a| a != ":")
+            .map(|arg| match find_newest(arg) {
+                Err(e) if e.kind() == ErrorKind::NotFound => Ok(SystemTime::UNIX_EPOCH),
+                n @ _ => n,
```
Comment (alecthomas) [inline src/main.rs:73]: Just when you think Rust can't throw more magical shit at you, it surprises you.

Comment (alecthomas) [reply src/main.rs:73]: The irony is that it's the same length as `Ok(n) => n`!

File: src/main.rs [hunk 6]
```diff
@@ -75,53 +63,35 @@ struct Newer {
...
-            Input,
-            Command,
-        }
+        let mut args = args.iter();
+
+        let newest = args
+            .by_ref()
+            .take_while(|&a| a != ":")
+            .map(|arg| match find_newest(arg) {
+                Err(e) if e.kind() == ErrorKind::NotFound => Ok(SystemTime::UNIX_EPOCH),
```
Comment (alecthomas) [inline src/main.rs:72]: This seems to be dropping the other errors?

File: src/main.rs [hunk 7]
```diff
@@ -75,53 +63,35 @@ struct Newer {
...
-        enum State {
-            Output,
-            Input,
-            Command,
-        }
+        let mut args = args.iter();
+
+        let newest = args
+            .by_ref()
+            .take_while(|&a| a != ":")
```
Comment (alecthomas) [inline src/main.rs:70]: This seems like it's no longer finding the newest outputs, only the newest inputs, or am I mistaken?

**Review by alecthomas — COMMENTED** [review]
I folded a few of these changes into main as I'd already made other changes, but I think for me it's clearer leaving it as a state machine rather than using iterators.

File: src/main.rs [hunk 8]
```diff
@@ -13,7 +13,7 @@ use tracing_subscriber::{fmt, prelude::*, EnvFilter};
 /// all subsequent arguments up to "--". If a "--" argument is present, all
 /// arguments after it are executed as a command if any input is newer than all
 /// outputs.
-fn main() {
+fn main() -> Result<(), Error> {
```
Comment (alecthomas) [inline src/main.rs:16]: This seems to be frowned upon by Google, preferring exit(). I would imagine one downside is that you have no control over the formatting.

Comment (quad) [reply src/main.rs:16]: - You can use `?` in `main`, the same as anywhere else
- The [`Termination` trait](https://doc.rust-lang.org/stable/std/process/trait.Termination.html) is stable and controls behaviour like formatting and exit codes

Use `fn main() -> ExitCode` and `return ExitCode::SUCCESS` / `return ExitCode::from(123)` / `return 123u8.into()`. IMHO calling _any_ function that doesn't return (exit, panic, …) is a smell.

Anyway, at least when I read it last year, my opinion of Google's Rust style guide was not high…

File: src/main.rs (hunk 5 above)
Comment (quad) [reply src/main.rs:73]: It's different. `n @ _ => n` matches all values. In this case, it replaces:

```rust
Err(e) => e,
Ok(v) => v,
```

File: src/main.rs (hunk 6 above)
Comment (quad) [reply src/main.rs:72]: [See my following comment.](https://github.com/alecthomas/mk/pull/3/files#r1531444064)

File: src/main.rs (hunk 7 above)
Comment (quad) [reply src/main.rs:70]: In short:
```rust
newest = inputs.map { |i| find_newest(i) }.max().unwrap_or(SystemTime::UNIX_EPOCH)
newers = outputs.map { |o| find_newest(o) }
newer = if newers.is_empty() {
    true
} else {
    newers.any(|n| n > newest)
};
```

That is:
1. `newest` is the timestamp of the newest input.
2. `newers` are the timestamps of the outputs.
3. `newer` is true if any of the `newers` are more recent then the `newest` or if there are no `newers`.

Did I misunderstand the original logic?

**Comment by quad** [comment]
Closed because @alecthomas reviewed and made changes out of band.

**Pull request closed without merging**